src/job.ts

```diff
@@ -181,6 +182,21 @@ export class Job {
         predefinedVariables["CI_REGISTRY"] = `local-registry.${this.gitData.remote.host}`;
         predefinedVariables["CI_REGISTRY_IMAGE"] = `$CI_REGISTRY/${this._variables["CI_PROJECT_PATH"].toLowerCase()}`;
 
+        // Expand variables in rules:changes
+        if (this.rules && expandVariables) {
+            const expanded = Utils.expandVariables(this._variables);
+            this.rules.forEach((rule, ruleIdx, rules) => {
+                const changes = Array.isArray(rule.changes) ? rule.changes : rule.changes?.paths;
+                if (!changes || !Array.isArray(changes)) {
```

Comment:
```suggestion
                if (!changes) {
```
what is `!Array.isArray(changes)` for?

Reply:
To guard against `rule.changes?.paths` not being an array.

Reply:
> since schema validation can be disabled
In my opinion, i think this can be removed,

Because on a typical(99% of users) scenario, schema validation should never be disabled, unless there's some bug with it. 

otherwise these type of defensive statement will eventually bloat this program in the future.